intermediate_source/torchserve_with_ipex_2.rst

```diff
@@ -0,0 +1,471 @@
+
+Let's look at Intel® VTune Profiler ITT trace. In the example script, we've annotated each *step_x* of the inference loop.
+
+.. figure:: /_static/img/torchserve-ipex-images-2/6.png
+   :width: 100%
+   :align: center
+   
+Each step is traced in the timeline graph. The duration of model inference on the last step (step_99) decreased from 304.308 ms to 261.843 ms. 
+
+Exercise with TorchServe
+^^^^^^^^^^^^^^^^^^^^^^^^
+Let's profile PTMalloc vs. JeMalloc with TorchServe.
+
+We'll use `TorchServe apache-bench benchmarking <https://github.com/pytorch/serve/tree/master/benchmarks#benchmarking-with-apache-bench>`_ with ResNet50 FP32, batch size 32, concurrency 32, requests 8960. All other parameters are the same as the `default parameters <https://github.com/pytorch/serve/tree/master/benchmarks#benchmark-parameters>`_. 
+
+As in the previous exercise, we'll use the launcher to designate the memory allocator, and to bind the workload to physical cores of the first socket. To do so, user simply needs to add a few lines in `config.properties <https://pytorch.org/serve/configuration.html#config-properties-file>`_: 
+
+PTMalloc
+
+.. code:: python
+
+    cpu_launcher_enable=true
+    cpu_launcher_args=--node_id 0 --use_default_allocator
+  
+JeMalloc 
+
+.. code:: python
+
+    cpu_launcher_enable=true
+    cpu_launcher_args=--node_id 0 --enable_jemalloc
+    
+Let's collect level-1 TMA metrics. 
+
+.. figure:: /_static/img/torchserve-ipex-images-2/7.png
+   :width: 100%
+   :align: center
+
+Let's go one level deeper. 
+
+.. figure:: /_static/img/torchserve-ipex-images-2/8.png
+   :width: 100%
+   :align: center
+
+Let's use Intel® VTune Profiler ITT to annotate `TorchServe inference scope <https://github.com/pytorch/serve/blob/master/ts/torch_handler/base_handler.py#L188>`_ to profile at inference-level granularity. As `TorchServe Architecture <https://github.com/pytorch/serve/blob/master/docs/internals.md#torchserve-architecture>`_ consists of several sub-components, including the Java frontend for handling request/response, and the Python backend for running the actual inference on the models, it is helpful to use Intel® VTune Profiler ITT to limit the collection of trace data at inference-level.  
+
+.. figure:: /_static/img/torchserve-ipex-images-2/9.png
+   :width: 100%
+   :align: center
+   
+Each inference call is traced in the timeline graph. The duration of the last model inference decreased from 561.688 ms to 251.287 ms - 2.2x speedup.
+
+.. figure:: /_static/img/torchserve-ipex-images-2/10.png
+   :width: 100%
+   :align: center
+   
+The timeline graph can be expanded to see op-level profiling results. The duration of *aten::conv2d* decreased from 16.401 ms to 6.392 ms - 2.6x speedup. 
+
+In this section, we've demonstrated that JeMalloc can give better performance than the default PyTorch memory allocator, PTMalloc, with efficient thread-local caches improving Back End Bound.
+
+Intel® Extension for PyTorch*
```

Comment:
This almost looks like a separate tutorial?

Reply:
It's part of this tutorial, a short introduction to Intel® Extension for PyTorch*, before jumping into TorchServe with Intel® Extension for PyTorch*